app/react/src/server/cra-config.ts

```diff
@@ -71,43 +71,53 @@ export function isReactScriptsInstalled(requiredVersion = '2.0.0') {
 }
 
 export const getRules = (extensions: string[]) => (rules: RuleSetRule[]) =>
-  rules.reduce((craRules: any, rule: any) => {
-    // If at least one extension satisfies the rule test, the rule is one
-    // we want to extract
-    if (rule.test && extensions.some(normalizeCondition(rule.test))) {
-      // If the base test is for extensions, return early
-      return craRules.concat(rule);
-    }
+  rules.reduce(
+    (craRules, rule) => {
+      // If at least one extension satisfies the rule test, the rule is one
+      // we want to extract
+      if (rule.test && extensions.some(normalizeCondition(rule.test))) {
+        // If the base test is for extensions, return early
+        return craRules.concat(rule);
+      }
 
-    // Get any rules contained in rule.oneOf
-    if (!rule.test && rule.oneOf) {
-      craRules.push(...getRules(extensions)(rule.oneOf));
-    }
+      // Get any rules contained in rule.oneOf
+      if (!rule.test && rule.oneOf) {
+        craRules.push(...getRules(extensions)(rule.oneOf));
+      }
 
-    // Get any rules contained in rule.rules
-    if (!rule.test && rule.rules) {
-      craRules.push(...getRules(extensions)(rule.rules));
-    }
+      // Get any rules contained in rule.rules
+      if (!rule.test && rule.rules) {
+        craRules.push(...getRules(extensions)(rule.rules));
+      }
 
-    return craRules;
-  }, []);
+      return craRules;
+    },
+    [] as RuleSetRule[]
+  );
 
 const getStyleRules = getRules(cssExtensions.concat(cssModuleExtensions));
 
 export const getTypeScriptRules = (webpackConfigRules: RuleSetRule[], configDir: string) => {
   const rules = getRules(typeScriptExtensions)(webpackConfigRules);
-  // We know CRA only has one rule targeting TS for now, which is the first rule.
-  const babelRule = rules[0];
-  // Resolves an issue where this config is parsed twice (#4903).
-  if (typeof babelRule.include !== 'string') return rules;
+
   // Adds support for using TypeScript in the `.storybook` (or config) folder.
-  return [
-    {
-      ...babelRule,
-      include: [babelRule.include, path.resolve(configDir)],
+  return rules.reduce(
+    (accRules, rule) => {
+      // Resolves an issue where this config is parsed twice (#4903).
+      if (typeof rule.include !== 'string') {
```

Comment:
i really dont like this `if` here. It says it fixes double parsing, but it also ignores loaders of CRA that have array in `include` prop. Currently there is no TSX loader with array in `include` but I am sure there should be a proper fix to avoid double parsing instead of this one

Reply:
We're in the process of rewriting this (it'll not make 5.2 though, which is why this PR is great).

I agree with what you say, this was a concession made previously after a discussion with @igor-dv.